ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/bridge/validate/DeduplicationState.scala

```diff
@@ -22,25 +21,28 @@ case class DeduplicationState private (
       commandDeduplicationDuration: Duration,
       recordTime: Time.Timestamp,
   ): (DeduplicationState, Boolean) = {
+    assert(
+      deduplicationQueue.lastRecordTimeOption.forall(_ <= recordTime),
+      s"Inserted record time ($recordTime) for changeId ($changeId) cannot be before the last inserted record time (${deduplicationQueue.lastRecordTimeOption}).",
+    )
+    assert(
+      commandDeduplicationDuration.compareTo(maxDeduplicationDuration) <= 0,
+      s"Cannot deduplicate for a period ($commandDeduplicationDuration) longer than the max deduplication duration ($maxDeduplicationDuration).",
+    )
+
     bridgeMetrics.SequencerState.deduplicationQueueLength.update(deduplicationQueue.size)
-    if (commandDeduplicationDuration.compareTo(maxDeduplicationDuration) > 0)
-      throw new RuntimeException(
-        s"Cannot deduplicate for a period ($commandDeduplicationDuration) longer than the max deduplication duration ($maxDeduplicationDuration)."
-      )
-    else {
-      val expiredTimestamp = expiredThreshold(maxDeduplicationDuration, recordTime)
 
-      val queueAfterEvictions = deduplicationQueue.dropWhile(_._2 <= expiredTimestamp)
+    val expiredTimestamp = expiredThreshold(maxDeduplicationDuration, recordTime)
+    val queueAfterEvictions = deduplicationQueue.withoutOlderThan(expiredTimestamp)
 
-      val isDuplicateChangeId = queueAfterEvictions
-        .get(changeId)
-        .exists(_ >= expiredThreshold(commandDeduplicationDuration, recordTime))
+    val isDuplicateChangeId = queueAfterEvictions
+      .get(changeId)
+      .exists(_ >= expiredThreshold(commandDeduplicationDuration, recordTime))
 
-      if (isDuplicateChangeId)
-        copy(deduplicationQueue = queueAfterEvictions) -> true
-      else
-        copy(deduplicationQueue = queueAfterEvictions.updated(changeId, recordTime)) -> false
-    }
+    if (isDuplicateChangeId)
+      copy(deduplicationQueue = queueAfterEvictions) -> true
```

Comment:
AFAICS if flag is true we throw away the state on client side. Why are we modifying it then? Why are we throwing it away above?

Reply:
This method also evicts. So even on duplicates we return a trimmed version of it (i.e. without the entries that have expired based on the current recordTime. Does this answer your question?

Reply:
yes, but we do not seem to use the trimmed version on the caller side

Reply:
Yes, good catch there. Fixed e0a354e086cdcfda2e09f0791f1d2babf9bc43e3